File: DATA/APP/NeuralLibrary/DataSet.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace NeuralLibrary
{
    /// <summary>
    /// Holds an abstract dataset.
    /// </summary>
    public abstract class DataSet : List<DataPoint>
    {
        /// <summary>
        /// Load dataset into the list.
        /// </summary>
        public abstract void Load();

        /// <summary>
        /// Random used for the dataset extension (shuffle)
        /// </summary>
        public static Random r = new Random();

        /// <summary>
        /// An extension for all ILists and dataset which shuffles the order.
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="list"></param>
        public DataSet Shuffle()
        {
            int n = this.Count;
            while (n > 1)
            {
                n--;
                int k = r.Next(n + 1);
                DataPoint value = this[k];
                this[k] = this[n];
                this[n] = value;
            }

            return this;
        }

        /// <summary>
        /// Calculates the errors based on the datapoints in a given dataset.
        /// </summary>
        /// <param name="nn">The network from which to calculate the errors</param>
        /// <param name="step">The step size.</param>
        /// <returns></returns>
        public double[] CalculateErrors(Network nn, double step = -1)
        {
            return this.Select((d) =>
                {
                    nn.FeedForward(d.Input);
                    return 0.5 * d.Desired.Select(
                        (x, i) => Math.Pow(Gaussian.Step(nn.Output[i], step) - x, 2)).Sum();
                }).ToArray();
        }

        public double CalculateError(Network nn, double step = -1)
        {
            return CalculateErrors(nn, step).Sum();
        }

    }
}

File: Python/Neural Network/dataset.py
import random
#list is a list of datapoints
def step(p, stepPoint):
    if (stepPoint == -1):
        return p
    if (p <= stepPoint):
        return 0
    else:
        return 1
class DataSet(list):
    def load(self):
        pass
    def shuffle(self):
        n = self.count
        while n > 1:
            n -= 1
            k = random.randint(0,n+1)
            value = self[k]
            self[k] = self[n]
            self[n] = value
        return self
    def calculateErrors(self, nn, step = -1):
        output = list()
        for d in self:
            nn.feedForward(d.input)
            sum = 0
            i = 0
            #WTF is (x,i)?!?!
            for desire in d.desired:
                i+=1
                sum+=(step(nn.Output[i], step)-i)**2
            output.append(0.5*sum)
            
        return 
    def totalError(self, nn, step = -1):
        output = 0;
        for error in self.calculateErrors(nn, step):
            output += step
        return output
    def toString(self):
        output ="\n"
        for d in self:
            output+=str(d.input)+"--->"+str(d.desired)
        return output